ably/realtime_conn_spec_test.go

```diff
@@ -562,3 +563,95 @@ func TestRealtimeConn_RTN15c3(t *testing.T) {
 		t.Errorf("expected %d got %d", errInfo.StatusCode, reason.StatusCode)
 	}
 }
+
+func TestRealtimeConn_RTN15c4(t *testing.T) {
+	t.Parallel()
+
+	doEOF := make(chan struct{}, 1)
+
+	type meta struct {
+		dial     *url.URL
+		messages []*proto.ProtocolMessage
+	}
+
+	var metaList []*meta
+	errInfo := &proto.ErrorInfo{
+		StatusCode: http.StatusBadRequest,
+	}
+	app, client := ablytest.NewRealtimeClient(&ably.ClientOptions{
+		NoConnect: true,
+		Dial: func(protocol string, u *url.URL) (proto.Conn, error) {
+			m := &meta{dial: u}
+			metaList = append(metaList, m)
+			c, err := ablyutil.DialWebsocket(protocol, u)
+			return protoConnWithFakeEOF{Conn: c, doEOF: doEOF, onMessage: func(msg *proto.ProtocolMessage) {
+				if len(metaList) == 2 && len(m.messages) == 0 {
+					msg.Action = proto.ActionError
+					msg.Error = errInfo
+				}
+				m.messages = append(m.messages, msg)
+			}}, err
+		},
+	})
+	defer safeclose(t, client, app)
+
+	if err := ablytest.Wait(client.Connection.Connect()); err != nil {
+		t.Fatalf("Connect=%s", err)
+	}
+
+	channel := client.Channels.Get("channel")
+	if err := ablytest.Wait(channel.Attach()); err != nil {
+		t.Fatal(err)
+	}
+	chanStateChanges := make(chan ably.State, 18)
+	channel.On(chanStateChanges, ably.StateChanFailed)
+
+	stateChanges := make(chan ably.State, 16)
+	client.Connection.On(stateChanges)
+
+	doEOF <- struct{}{}
+
+	var state ably.State
+
+	select {
+	case state = <-stateChanges:
+	case <-time.After(50 * time.Millisecond):
+		t.Fatal("didn't transition on EOF")
+	}
+	if expected, got := ably.StateConnDisconnected, state; expected != got.State {
+		t.Fatalf("expected transition to %v, got %v", expected, got)
+	}
+	rest, err := ably.NewRestClient(app.Options())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = rest.Channels.Get("channel", nil).Publish("name", "data")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case state = <-stateChanges:
+	case <-time.After(50 * time.Millisecond):
+		t.Fatal("didn't reconnect")
+	}
+	if expected, got := ably.StateConnConnecting, state; expected != got.State {
+		t.Fatalf("expected transition to %v, got %v", expected, got)
+	}
+	select {
+	case state = <-chanStateChanges:
+	case <-time.After(80 * time.Millisecond):
+		t.Fatal("didn't change state")
+	}
+	if expected, got := ably.StateChanFailed, state; expected != got.State {
+		t.Fatalf("expected transition to %v, got %v", expected, got)
+	}
+	err = client.Connection.Reason()
+	if err == nil {
+		t.Fatal("expected reason to be set")
+	}
+	reason := err.(*ably.Error)
+	if reason.StatusCode != errInfo.StatusCode {
+		t.Errorf("expected %d got %d", errInfo.StatusCode, reason.StatusCode)
+	}
```

Comment:
Missing this:

> The client should transition to the FAILED state

(Here "client" means the connection.)

Reply:
addressed 402872e3efa6fcf4b3eea62a349a58ffe1932fa5